sync/src/types/header_map/backend_sled.rs

```diff
@@ -0,0 +1,104 @@
+use super::KeyValueBackend;
+use crate::types::HeaderView;
+use ckb_types::{packed::Byte32, prelude::*};
+use sled::Db;
+use std::path;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use tempfile::TempDir;
+
+pub(crate) struct SledBackend {
+    count: AtomicUsize,
+    db: Db,
+    _tmpdir: TempDir,
+}
+
+impl KeyValueBackend for SledBackend {
+    fn new<P>(tmp_path: Option<P>) -> Self
+    where
+        P: AsRef<path::Path>,
+    {
+        let mut builder = tempfile::Builder::new();
+        builder.prefix("ckb-tmp-");
+        let tmpdir = if let Some(ref path) = tmp_path {
+            builder.tempdir_in(path)
+        } else {
+            builder.tempdir()
+        }
+        .expect("failed to create a tempdir to save header map into disk");
+
+        let db: Db = sled::open(tmpdir.path())
+            .expect("failed to open a key-value database to save header map into disk");
+        Self {
+            db,
+            _tmpdir: tmpdir,
+            count: AtomicUsize::new(0),
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.count.load(Ordering::SeqCst)
+    }
+
+    fn contains_key(&self, key: &Byte32) -> bool {
+        self.db
+            .contains_key(key.as_slice())
+            .expect("sled contains_key")
+    }
+
+    fn get(&self, key: &Byte32) -> Option<HeaderView> {
+        self.db
+            .get(key.as_slice())
+            .unwrap_or_else(|err| panic!("read header map from disk should be ok, but {}", err))
+            .map(|slice| HeaderView::from_slice_should_be_ok(slice.as_ref()))
+    }
+
+    fn insert(&self, value: &HeaderView) -> Option<()> {
+        let key = value.hash();
+        let last_value = self
+            .db
+            .insert(key.as_slice(), value.to_vec())
+            .expect("failed to insert item to sled");
+        if last_value.is_none() {
+            self.count.fetch_add(1, Ordering::SeqCst);
+        }
+        last_value.map(|_| ())
+    }
+
+    fn insert_batch(&self, values: &[HeaderView]) {
+        let mut count = 0;
+        for value in values {
+            let key = value.hash();
+            let last_value = self
```

Comment:
Have you compared the performance difference between consecutive insertions with [`apply_batch`](https://docs.rs/sled/latest/sled/struct.Db.html#method.apply_batch)?

Reply:
[apply_batch](https://docs.rs/sled/latest/sled/struct.Db.html#method.apply_batch) has the same problem, it can't give feedback whether a key was overwritten or written directly，Here also tried to compare with [transaction](https://docs.rs/sled/latest/sled/transaction/index.html)，transaction performs worse because of the extra overhead required to guarantee ACID.